### PR TITLE
fix(platform): stub torch.distributed._symmetric_memory for torch < 2.8

### LINKED ISSUE #509: fix(platform): stub torch.distributed._symmetric_memory for torch < 2.8
Release tracking issue for the `flagos2.2-rc2` release.

Linked PR: https://github.com/flagos-ai/vllm-plugin-FL/pull/491

`vllm.distributed.parallel_state` imports `torch.distributed._symmetric_memory` at module level, but that module only exists in PyTorch 2.8+. Vendor torch builds on 2.7.x — iluvatar corex ships torch 2.7.1 — therefore crash with `ImportError` before the engine core starts, in both the API-server and spawned worker processes.

### PR DESCRIPTION
**Pull request by tengqm**
Fixes #509

## Summary

`vllm.distributed.parallel_state` imports `torch.distributed._symmetric_memory`
at module level, but that module only exists in PyTorch 2.8+. Vendor torch
builds on 2.7.x — iluvatar corex ships torch 2.7.1 — therefore crash with
`ImportError` before the engine core starts, in both the API-server and spawned
worker processes.

This PR pre-registers an empty stub module (plus a mirror of the name on the
`torch.distributed` parent package, since a bare `sys.modules` entry does not
set the parent attribute) whenever `torch.distributed._symmetric_memory` is
absent. All of vllm's runtime uses of the module are lazy (inside functions),
so the stub only needs to satisfy the module-level import gate; it is never
consulted on the verified greedy / sampling paths. The probe imports the module
and registers the stub only when that import fails with a `ModuleNotFoundError`
naming the module itself, so PyTorch 2.8+ builds get the real module and a
dependency missing from a module that does exist still surfaces (see Review
follow-up).

The probe stays capability-based at module top, like the existing
`torch.float4_e2m1fn_x2` sentinel in the same file.

## Notes

- Root-caused on-node: iluvatar-corex4.4.0 (torch 2.7.1) with the vllm 0.24.0
  app image, serve-gate crash at `vllm/distributed/parallel_state.py:42`.
  A container-level `try/except` around the import made serve fully green
  (greedy decode and temperature>0 sampling both clean).
- Deliberately not gated on a specific vendor: any torch build < 2.8 lacks the
  module, and on 2.8+ the real module is imported untouched.

## Review follow-up

`hasattr(torch.distributed, "_symmetric_memory")` read False on 2.8+ too —
`torch.distributed` never imports the submodule eagerly, every reference in
torch is lazy — so the stub was registered on every torch version and shadowed
the real module for the rest of the process. The corex4.5.0 rows below passed
because symm mem is opt-in (`VLLM_ALLREDUCE_USE_SYMM_MEM`), so nothing
consulted the shadowed name.

The probe is now `importlib.import_module("torch.distributed._symmetric_memory")`
and the stub is registered only on a `ModuleNotFoundError` whose `.name` is that
module. A/B against fake torch trees, previous head vs this head:

| tree | previous head | this head |
|---|---|---|
| module absent (2.7 layout) | stub | stub |
| module present (2.8 layout) | stub shadows the real module | real module kept |
| real module, missing dependency | stub, error swallowed | `ModuleNotFoundError` propagates |

## Verification

The wheel built from the previous head (`0.2.1+gc9e2573.d20260913`) was installed
into both iluvatar corex vllm 0.24.0 app images and verified end-to-end:

| Backend | Compiler | Serve ready | Completion |
|---|---|---|---|
| iluvatar-corex4.4.0 (torch 2.7.1) | flagtree | ~105s | coherent |
| iluvatar-corex4.4.0 (torch 2.7.1) | vendor triton 3.1 | — | not deliverable, vendor compiler defect (unrelated to this change) |
| iluvatar-corex4.5.0 (torch 2.10) | flagtree | ~90s | clean |
| iluvatar-corex4.5.0 (torch 2.10) | vendor triton 3.2 | ~150s | clean |

The corex4.5.0 rows neither confirm nor refute the stub's effect: they ran to
completion with the stub registered, which is what a shadowed-but-unconsulted
module looks like, not what an untouched real module looks like. With the probe
fixed, both corex4.5.0 rows would exercise the real module for the first time.

Node verification was not re-run for the current head. The wheel that was
verified (`0.2.1+gc9e2573.d20260913`) is built from the previous head; the
change since is confined to the probe block, whose < 2.8 branch reaches the same
statements as before, and whose 2.8+ branch restores the pre-PR behaviour
(no stub at all). The A/B table above is the evidence for this head.

---

This PR was written in part with the assistance of generative AI.

### CODE REVIEW

**Review by ceci3 — APPROVED** [review]

File: vllm_fl/__init__.py [hunk 1]
```diff
@@ -14,10 +14,30 @@
...
+# vllm.distributed.parallel_state imports it at module level, so vendor
+# torch builds < 2.8 (iluvatar corex on 2.7.x) die with ImportError before
+# engine start. vllm's later uses of the module are lazy, so an empty
+# pre-registered stub suffices to get past the import gate.
+import torch.distributed as _torch_distributed
+
+if not hasattr(_torch_distributed, "_symmetric_memory"):
+    import types as _types
+
+    _symm_mem_stub = _types.ModuleType("torch.distributed._symmetric_memory")
```
Comment (ceci3) [inline vllm_fl/__init__.py:32]: hasattr(torch.distributed, "_symmetric_memory") only checks whether the submodule has already been imported; it can be false even when the real module exists. Please try importlib.import_module("torch.distributed._symmetric_memory") and create the stub only when that module itself raises ModuleNotFoundError, checking exc.name to avoid masking a missing dependency.

Comment (tengqm) [reply vllm_fl/__init__.py]: Fixed in c2b0714 — this was a live bug rather than a theoretical one. `torch.distributed.__init__` never pulls the submodule in (every `_symmetric_memory` reference in torch is lazy, inside functions), so `hasattr()` read False on 2.8+ too and the stub was registered there as well, shadowing the real module for the rest of the process. The corex4.5.0 rows in the table passed because symm mem is opt-in (`VLLM_ALLREDUCE_USE_SYMM_MEM`), so nothing consulted the shadowed name.

The probe is now `importlib.import_module("torch.distributed._symmetric_memory")`, and the stub is registered only when that raises `ModuleNotFoundError` whose `.name` is that module — a missing dependency of a module that does exist keeps raising. A/B against fake torch trees, previous head vs c2b0714:

| tree | previous head | c2b0714 |
|---|---|---|
| module absent (2.7 layout) | stub | stub |
| module present (2.8 layout) | stub shadows the real module | real module kept |
| real module, missing dependency | stub, error swallowed | `ModuleNotFoundError` propagates |
